lib/eqaf.ml

```diff
@@ -111,3 +111,67 @@ let equal a b =
   if (al lxor ln) lor (bl lxor ln) <> 0
   then false
   else equal ~ln a b
+
+let[@inline always] compare (a:int) b = a - b
+let[@inline always] minus_one_or_less n = ((n land min_int) asr Sys.int_size) land 1
+let[@inline always] one_if_not_zero n = (minus_one_or_less n) lor (minus_one_or_less (-n))
+
+let compare_le ~ln a b =
+  let r = ref 0 in
+  let i = ref (pred ln) in
+
+  while !i >= 0 do
+    let xa = get a !i and xb = get b !i in
+    let c = compare xa xb in
+    let n = minus_one_or_less c + one_if_not_zero c in
+    let n = n lsr ((one_if_not_zero !r) lsl 1) in
+    r := n + !r ;
+    decr i ;
+  done ;
+
+  (!r land 1) - (!r lsr 1)
+
+let compare_le_with_len ~len:ln a b =
+  let al = String.length a in
+  let bl = String.length b in
+  if ln <= al && ln <= bl
```

Comment:
this will short-circuit (`&&`) and thus the cases where `ln <= al && ln > bl` or `ln > al && ln <= bl` are distinguishable.